account-kit/react/src/hooks/useUiConfig.tsx

```diff
@@ -0,0 +1,129 @@
+"use client";
+
+import {
+  createContext,
+  useContext,
+  useEffect,
+  useRef,
+  type PropsWithChildren,
+} from "react";
+import { create, useStore, type StoreApi } from "zustand";
+import { useShallow } from "zustand/react/shallow";
+import { IS_SIGNUP_QP } from "../components/constants.js";
+import type {
+  AlchemyAccountsUIConfig,
+  AuthIllustrationStyle,
+} from "../types.js";
+import { useSignerStatus } from "./useSignerStatus.js";
+
+type AlchemyAccountsUIConfigWithDefaults = Omit<
+  Required<AlchemyAccountsUIConfig>,
+  "auth"
+> & {
+  auth: NonNullable<Required<AlchemyAccountsUIConfig["auth"]>>;
+};
+
+export type UiConfigStore = AlchemyAccountsUIConfig & {
+  isModalOpen: boolean;
+  setModalOpen: (isOpen: boolean) => void;
+  updateConfig: (config: AlchemyAccountsUIConfig) => void;
+};
+
+export const DEFAULT_UI_CONFIG: AlchemyAccountsUIConfigWithDefaults = {
+  illustrationStyle: "flat" as AuthIllustrationStyle,
+  auth: {
+    addPasskeyOnSignup: false,
+    header: null,
+    hideError: false,
+    sections: [[{ type: "email" }], [{ type: "passkey" }]],
+    onAuthSuccess: () => {},
+    hideSignInText: false,
+  },
+  modalBaseClassName: "",
+  supportUrl: "",
+};
+
+export function createUiConfigStore(
+  initialConfig: AlchemyAccountsUIConfig = DEFAULT_UI_CONFIG
+) {
+  return create<UiConfigStore>((set) => ({
+    ...initialConfig,
+    isModalOpen: false,
+    setModalOpen: (isOpen: boolean) => {
+      set({ isModalOpen: isOpen });
+    },
+    updateConfig: (config: AlchemyAccountsUIConfig) => {
+      set(() => ({ ...config }));
+    },
+  }));
+}
+
+const UiConfigContext = createContext<StoreApi<UiConfigStore> | undefined>(
+  undefined
+);
+
+export function useUiConfig<T extends unknown | undefined = UiConfigStore>(
```

Comment:
Does `extends unknown | undefined` do anything? All types should extend `unknown | undefined`, but maybe this helps somehow with error messages or autocomplete?

Reply:
I think that works fine with just `useUiConfig<T = UiConfigStore>`. This seems to be behaving: [Playground](https://www.typescriptlang.org/play/?#code/JYOwLgpgTgZghgYwgAgKrAMIHsQ2AcwGUwsoUBvAWAChlkBXYALmQGcwpR8BuGgXxo0Y9EAjDAcDVhHTZcBADwAVZAF40mHHiIkyAPgAUNOnWkAbCGNIB+FgfZxILWVoLFSEAJRq9yJTU8WFSpaZDAACygsAHdkEAhYgFEoKKgDACIAOSwwZGAAWwAHC3yIcAgAE3TPfkFqBBx2ZAb5fDUpGU1Wg09eahoWpsZ2+mkXboNyBmBkPm9VX0ZeoA)

Reply:
oh yea neat that's much nicer here and eliminates the need for overloading which is super hard to read I find

Reply:
but it does work if I keep the function overloading